src/webgpu/shader/execution/robust_access.spec.ts

```diff
@@ -0,0 +1,426 @@
+
+  // Contains just zero (for now).
+  const constantsBuffer = t.device.createBuffer({ size: 4, usage: GPUBufferUsage.UNIFORM });
+
+  const resultBuffer = t.device.createBuffer({
+    size: 4,
+    usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.STORAGE,
+  });
+
+  const source = `
+    [[block]] struct Constants {
+      zero: u32;
+    };
+    [[group(1), binding(0)]] var<uniform> constants: Constants;
+
+    [[block]] struct Result {
+      value: u32;
+    };
+    [[group(1), binding(1)]] var<storage, write> result: Result;
+
+    ${testSource}
+
+    [[stage(compute)]]
+    fn main() {
+      ignore(constants.zero); // Ensure constants buffer is statically-accessed
+      result.value = runTest();
+    }`;
+
+  t.debug(source);
+  const module = t.device.createShaderModule({ code: source });
+  const pipeline = t.device.createComputePipeline({
+    compute: { module, entryPoint: 'main' },
+  });
+
+  const group = t.device.createBindGroup({
+    layout: pipeline.getBindGroupLayout(1),
+    entries: [
+      { binding: 0, resource: { buffer: constantsBuffer } },
+      { binding: 1, resource: { buffer: resultBuffer } },
+    ],
+  });
+
+  const testGroup = t.device.createBindGroup({
+    layout: pipeline.getBindGroupLayout(0),
+    entries: testBindings,
+  });
+
+  const encoder = t.device.createCommandEncoder();
+  const pass = encoder.beginComputePass();
+  pass.setPipeline(pipeline);
+  pass.setBindGroup(0, testGroup);
+  pass.setBindGroup(1, group);
+  pass.dispatch(1);
+  pass.endPass();
+
+  t.queue.submit([encoder.finish()]);
+
+  t.expectGPUBufferValuesEqual(resultBuffer, new Uint32Array([0]));
+}
+
+/** Fill an ArrayBuffer with sentinel values, except clear a region to zero. */
+function testFillArrayBuffer(
+  array: ArrayBuffer,
+  type: 'u32' | 'i32' | 'f32',
+  { zeroByteStart, zeroByteCount }: { zeroByteStart: number; zeroByteCount: number }
+) {
+  const constructor = { u32: Uint32Array, i32: Int32Array, f32: Float32Array }[type];
+  assert(zeroByteCount % constructor.BYTES_PER_ELEMENT === 0);
+  new constructor(array).fill(42);
+  new constructor(array, zeroByteStart, zeroByteCount / constructor.BYTES_PER_ELEMENT).fill(0);
+}
+
+const kArrayLength = 3;
+
+/**
+ * Generate a bunch of indexable types (vec, mat, sized/unsized array) for testing.
+ */
+function* arrayishTypes({
+  storageClass,
+  baseType,
+  atomic = false,
+}: {
+  /** Unsized array will only be generated for storageClass "storage". */
+  storageClass: string;
+  /** Base scalar type (i32/u32/f32/bool). */
+  baseType: ScalarType;
+  /** Whether to wrap the baseType in `atomic<>`. */
+  atomic?: boolean;
+}) {
+  const scalarInfo = kScalarTypeInfo[baseType];
+  if (atomic) assert(scalarInfo.supportsAtomics, 'type does not support atomics');
+  const scalarType = atomic ? `atomic<${baseType}>` : baseType;
+
+  // Vector types
+  for (const vectorType of kVectorContainerTypes) {
+    yield {
+      type: `${vectorType}<${scalarType}>`,
```

Comment:
Are you allowed to do `vec4<atomic<f32>>`??

Reply:
I'm not sure! Tint doesn't stop me, at least. I figure we'll revise the test if this turns out to be not allowed.

Reply:
I found that Tint also crashes in the same way on a clearly wrong thing, `vec2<vec2<>>`, so I'm going to assume it's not supposed to be allowed and take it out.